src/mpl/src/sock/mpl_sockaddr.c

```diff
@@ -0,0 +1,316 @@
+#include <stdio.h>
+
+#include "mpl_sockaddr.h"
+
+static int is_localhost(struct sockaddr *p_addr);
+
+static int af_type = AF_INET;
+static int _use_loopback = 0;
+static int _max_conn = SOMAXCONN;
+
+int MPL_get_sockaddr(const char *s_hostname, struct sockaddr_storage *p_addr)
+{
+    struct addrinfo ai_hint;
+    struct addrinfo *ai_list;
+    int ret;
+
+    memset(p_addr, 0, sizeof(*p_addr));
+    memset(&ai_hint, 0, sizeof(ai_hint));
+    ai_hint.ai_family = af_type;
+    ai_hint.ai_socktype = SOCK_STREAM;
+    ai_hint.ai_protocol = IPPROTO_TCP;
+    ai_hint.ai_flags = AI_ADDRCONFIG | AI_V4MAPPED;
+    ret = getaddrinfo(s_hostname, NULL, &ai_hint, &ai_list);
+    if (ret) {
+        return ret;
+    }
+    ret = 0;
+    if (af_type == AF_INET) {
+        memcpy(p_addr, ai_list->ai_addr, sizeof(struct sockaddr_in));
+    } else if (af_type == AF_INET6) {
+        memcpy(p_addr, ai_list->ai_addr, sizeof(struct sockaddr_in6));
+    } else {
+        ret = 1;
+    }
+    freeaddrinfo(ai_list);
+    return ret;
+}
+
+int MPL_get_sockaddr_direct(int type, struct sockaddr_storage *p_addr)
+{
+    memset(p_addr, 0, sizeof(*p_addr));
+    if (type < 0 || type > 1) {
+        return 1;
+    }
+    if (af_type == AF_INET) {
+        struct sockaddr_in *p_addr4 = (struct sockaddr_in *) p_addr;
+
+        p_addr4->sin_family = AF_INET;
+        if (type == 0) {
+            p_addr4->sin_addr.s_addr = htonl(INADDR_ANY);
+        } else if (type == 1) {
+            p_addr4->sin_addr.s_addr = htonl(0x7f000001);
+        }
+        return 0;
+    } else if (af_type == AF_INET6) {
+        struct sockaddr_in6 *p_addr6 = (struct sockaddr_in6 *) p_addr;
+
+        p_addr6->sin6_family = AF_INET6;
+        if (type == 0) {
+            p_addr6->sin6_addr = in6addr_any;
+        } else if (type == 1) {
+            p_addr6->sin6_addr = in6addr_loopback;
+        }
+        return 0;
+    } else {
+        return 1;
+    }
+}
+
+int MPL_get_sockaddr_iface(const char *s_iface, struct sockaddr_storage *p_addr)
+{
+    struct ifaddrs *ifaddr;
+    int ret;
+    struct ifaddrs *ifa;
+    int found = 0;
+
+    memset(p_addr, 0, sizeof(*p_addr));
+    ret = getifaddrs(&ifaddr);
+    if (ret) {
+        return ret;
+    }
+    ifa = ifaddr;
+    while (ifa) {
+        if (s_iface && ifa->ifa_name && strcmp(s_iface, ifa->ifa_name) != 0) {
+            ifa = ifa->ifa_next;
+            continue;
+        }
+        if (ifa->ifa_addr && ifa->ifa_addr->sa_family == af_type) {
+            found++;
+            if (af_type == AF_INET) {
+                memcpy(p_addr, ifa->ifa_addr, sizeof(struct sockaddr_in));
+            } else if (af_type == AF_INET6) {
+                memcpy(p_addr, ifa->ifa_addr, sizeof(struct sockaddr_in6));
+            }
+            if (!is_localhost((struct sockaddr *) ifa->ifa_addr)) {
+                break;
+            }
+        }
+        ifa = ifa->ifa_next;
+    }
+    freeifaddrs(ifaddr);
+    if (!found) {
+        return -1;
```

Comment:
The return value needs to be defined as an enum constant.

Reply:
The idea is to serve as wrapper and mimic the behavior of getifaddr, as well as the rest of the posix socket convention -- that return -1 on error and application checks errno for details.

Reply:
OK. I see. Then, I have no problem